README.md
# Animated Rating Bar
[![codersangam - animated_rating_bar](https://img.shields.io/static/v1?label=codersangam&message=animated_rating_bar&color=green&logo=github)](https://github.com/codersangam/animated_rating_bar "Go to GitHub repo")
[![stars - animated_rating_bar](https://img.shields.io/github/stars/codersangam/animated_rating_bar?style=social)](https://github.com/codersangam/animated_rating_bar)
[![forks - animated_rating_bar](https://img.shields.io/github/forks/codersangam/animated_rating_bar?style=social)](https://github.com/codersangam/animated_rating_bar)
[![GitHub release](https://img.shields.io/github/release/codersangam/animated_rating_bar?include_prereleases=&sort=semver&color=purple)](https://github.com/codersangam/animated_rating_bar/releases/)
[![License](https://img.shields.io/badge/License-MIT-purple)](#license)
[![issues - animated_rating_bar](https://img.shields.io/github/issues/codersangam/animated_rating_bar)](https://github.com/codersangam/animated_rating_bar/issues)

## Intro
A customizable animated rating bar for flutter.
## Getting started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  ...
  animated_rating_bar: <latest_version>
```

In your library add the following import:

```dart
import 'package:animated_rating_bar/animated_rating_bar.dart';
```

## Demo
![DEMO1](demo1.gif)
![DEMO2](demo2.gif)
![DEMO3](demo3.gif)
![DEMO4](dark_mode.gif)


## Special Thanks To Colleagues & 
1. [Rive](https://rive.app/community/3145-6649-star-rating/)
2. [Guidorosso](https://rive.app/@guidorosso/)

## License

Released under [MIT](/LICENSE) by [@codersangam](https://github.com/codersangam).

## Contribution

1. Fork it (https://github.com/codersangam/animated_rating_bar/fork)
2. Create your feature branch (git checkout -b feature/fooBar)
3. Commit your changes (git commit -am 'Add some fooBar')
4. Push to the branch (git push origin feature/fooBar)
5. Create a new Pull Request


CHANGELOG.md
## 0.0.8

* Animation added to normal rating bar.
* Demo Modified.
* Fixed Assets not found.

example/lib/test_screen.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:rive/rive.dart';

class RatingScreen extends StatefulWidget {
  const RatingScreen({super.key});

  @override
  State<RatingScreen> createState() => _RatingScreenState();
}

class _RatingScreenState extends State<RatingScreen> {
  StateMachineController? stateMachineController;
  SMIInput<double>? initialValue;

  double ratingValue = 0.0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Animated Rating Bar"),
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
      ),
      body: Column(
        children: [
          const SizedBox(height: 100),
          SizedBox(
            height: 40,
            width: 200,
            child: RiveAnimation.asset(
              "assets/rating_animation.riv",
              onInit: (artboard) {
                stateMachineController = StateMachineController.fromArtboard(
                  artboard,
                  "State Machine 1",
                  onStateChange: (stateMachineName, stateName) {
                    // print("$stateMachineName, $stateName");
                  },
                );
                if (stateMachineController != null) {
                  artboard.addController(stateMachineController!);
                  initialValue = stateMachineController!.findInput("Rating");
                  initialValue!.change(1);
                }

                artboard.forEachComponent(
                  (child) {
                    if (child is Shape) {
                      final Shape shape = child;
                      if (shape.name == "Star_base_1") {
                        debugPrint("Shape Name====> ${shape.name}");
                        debugPrint(
                            "Shape Fills====> ${shape.fills.first.children[0]}");
                        debugPrint(
                            "Shape Strokes====> ${shape.strokes.elementAt(0).paint.color}");
                      }

                      if (shape.name == "Star_1") {
                        (shape.fills.first.children[0] as SolidColor)
                            .colorValue = const Color(0xFFF44336).value;
                      } else if (shape.name == "Star_base_1") {
                        (shape.strokes.first.children[0] as SolidColor)
                            .colorValue = const Color(0xFFF44336).value;
                      } else if (shape.name == "Star_2") {
                        (shape.fills.first.children[0] as SolidColor)
                            .colorValue = const Color(0xFFF4E736).value;
                      }
                    }
                  },
                );
              },
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextField(
              onChanged: (value) {
                initialValue?.change(double.parse(value == "" ? "1" : value));
              },
            ),
          ),
          const SizedBox(height: 100),
          const Text("Rating Value"),
          Text(
            ratingValue.toString(),
            style: const TextStyle(fontSize: 30, fontWeight: FontWeight.bold),
          ),
          const SizedBox(height: 20),
          ElevatedButton(
            onPressed: () {
              int? id = initialValue!.id;
              if (kDebugMode) {
                print(stateMachineController!.getInputValue(id));
                setState(() {
                  ratingValue = stateMachineController!.getInputValue(id);
                });
              }
            },
            child: const Text("Get Value"),
          ),
        ],
      ),
    );
  }
}


test/animated_rating_bar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:animated_rating_bar/animated_rating_bar.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


example/lib/main.dart
import 'package:animated_rating_bar/animated_rating_bar.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Animated Rating Bar',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      darkTheme: ThemeData(useMaterial3: true, brightness: Brightness.dark),
      home: const RatingScreen(),
    );
  }
}

class RatingScreen extends StatelessWidget {
  const RatingScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: const Text("Animated Rating Bar"),
      ),
      body: Center(
        child: AnimatedRatingBar(
          activeFillColor: Theme.of(context).colorScheme.inversePrimary,
          strokeColor: Colors.green,
          initialRating: 0,
          height: 60,
          width: MediaQuery.of(context).size.width,
          animationColor: Colors.red,
          onRatingUpdate: (rating) {
            debugPrint(rating.toString());
          },
        ),
      ),
    );
  }
}


example/lib/rating_screen.dart
import 'package:animated_rating_bar/animated_rating_bar.dart';
import 'package:flutter/material.dart';

class RatingScreen extends StatelessWidget {
  const RatingScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: const Text("Lol"),
      ),
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          const SizedBox(height: 100),
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              // const Text("Lol"),
              AnimatedRatingBar(
                activeFillColor: Colors.red,
                strokeColor: Colors.red,
                // initialRating: 0,
                animationColor: Colors.green,
                onRatingUpdate: (rating) {
                  debugPrint(rating.toString());
                },
              ),
            ],
          ),
        ],
      ),
    );
  }
}


lib/widgets/animated_rating_bar.dart
import 'package:flutter/material.dart';
import 'package:rive/rive.dart';

class AnimatedRatingBar extends StatefulWidget {
  /// [AnimatedRatingBar] can be used for rating which enhaces User Experince
  ///
  /// by displaying beautiful animation like glows and sparks.
  ///
  /// Exmple:
  /// ```dart
  /// return AnimatedRatingBar(
  ///          activeFillColor: Theme.of(context).colorScheme.inversePrimary,
  ///          strokeColor: Theme.of(context).colorScheme.inversePrimary,
  ///          initialRating: 0,
  ///          height: 60,
  ///          width: MediaQuery.of(context).size.width,
  ///          animationColor: Theme.of(context).colorScheme.inversePrimary,
  ///          onRatingUpdate: (rating) {
  ///            debugPrint(rating.toString());
  ///          },
  ///        );
  /// ```
  const AnimatedRatingBar(
      {super.key,
      this.initialRating = 0.0,
      this.height,
      this.width,
      this.activeFillColor,
      this.strokeColor,
      required this.onRatingUpdate,
      this.animationColor});

  /// This sets Initial Rating of the Animated Rating Bar
  ///
  /// If not provided, by default it will start from 0.0
  final double? initialRating;

  /// Holds the height of the widget. You can customise it
  ///
  /// accoring to your requirements.
  final double? height;

  /// Holds the width of the widget. You can customise it
  ///
  /// accoring to your requirements.
  final double? width;

  /// Fills color on inner layer of icon except stroke.
  final Color? activeFillColor;

  /// You can even modify stroke color using this property.
  final Color? strokeColor;

  /// Animation color holds both glow and sparks color
  ///
  /// Use it accordingly.
  final Color? animationColor;

  /// This holds double value on updation of the rating.
  final ValueChanged<double> onRatingUpdate;

  @override
  State<AnimatedRatingBar> createState() => _AnimatedRatingBarState();
}

class _AnimatedRatingBarState extends State<AnimatedRatingBar> {
  StateMachineController? stateMachineController;
  SMIInput<double>? initialValue;
  double ratingValue = 0.0;

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: widget.height ?? 40,
      width: widget.width ?? 140,
      child: RiveAnimation.asset(
        "packages/animated_rating_bar/assets/new_rating_animation.riv",
        onInit: (artboard) {
          stateMachineController = StateMachineController.fromArtboard(
            artboard,
            "State Machine 1",
            onStateChange: (stateMachineName, stateName) {
              // print("$stateMachineName, $stateName");
              setState(() {
                int? id = initialValue!.id;
                ratingValue = stateMachineController!.getInputValue(id);
                widget.onRatingUpdate(ratingValue);
              });
            },
          );
          if (stateMachineController != null) {
            artboard.addController(stateMachineController!);
            initialValue = stateMachineController!.findInput("Rating");
            initialValue!.change(widget.initialRating ?? 1);
          }

          /// Customisation of each components from rive.
          artboard.forEachComponent(
            (child) {
              if (child is Shape) {
                final Shape shape = child;
                // if (shape.name == "Star_base_1") {
                //   debugPrint("Shape Name====> ${shape.name}");
                //   debugPrint(
                //       "Shape Fills====> ${shape.fills.first.children[0]}");
                //   debugPrint(
                //       "Shape Strokes====> ${shape.strokes.elementAt(0).paint.color}");
                // }
                if (widget.activeFillColor != null) {
                  if (shape.name == "Star_1" ||
                      shape.name == "Star_2" ||
                      shape.name == "Star_3" ||
                      shape.name == "Star_4" ||
                      shape.name == "Star_5") {
                    (shape.fills.first.children[0] as SolidColor).colorValue =
                        widget.activeFillColor!.value;
                  } else if (shape.name == "Star_base_1" ||
                      shape.name == "Star_base_2" ||
                      shape.name == "Star_base_3" ||
                      shape.name == "Star_base_4" ||
                      shape.name == "Star_base_5") {
                    (shape.strokes.first.children[0] as SolidColor).colorValue =
                        widget.strokeColor?.value ??
                            widget.activeFillColor!.value;
                  } else if (shape.name == "Star_4_glow" ||
                      shape.name == "Star_5_glow" ||
                      shape.name == "Star_5_sparks") {
                    (shape.strokes.first.children[0] as SolidColor).colorValue =
                        widget.animationColor?.value ??
                            widget.activeFillColor!.value;
                  }
                } else {
                  var brightness = MediaQuery.of(context).platformBrightness;
                  bool isDarkMode = brightness == Brightness.dark;
                  if (shape.name == "Star_1" ||
                      shape.name == "Star_2" ||
                      shape.name == "Star_3" ||
                      shape.name == "Star_4" ||
                      shape.name == "Star_5") {
                    (shape.fills.first.children[0] as SolidColor).colorValue =
                        isDarkMode
                            ? ThemeData.light().primaryColor.value
                            : ThemeData.dark().primaryColor.value;
                  } else if (shape.name == "Star_4_glow" ||
                      shape.name == "Star_5_glow" ||
                      shape.name == "Star_5_sparks") {
                    (shape.strokes.first.children[0] as SolidColor).colorValue =
                        isDarkMode
                            ? ThemeData.light().primaryColor.value
                            : ThemeData.dark().primaryColor.value;
                  }
                }
              }
            },
          );
        },
      ),
    );
  }
}
